pep-9999.rst

```diff
@@ -0,0 +1,247 @@
+PEP: TBD
+Title: Augmented Assignment Expressions
+Author: Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 14-May-2018
+Python-Version: 3.8
+Post-History: 14-May-2018
+
+
+Abstract
+========
+
+This is a proposal to allow augmented assignment statements to be used as
+expressions when the assignment target is a simple name.
+
+Associated changes are proposed to augmented assignment scoping rules to
+make them easier to use in nested functions, including those implicitly
+created as part of generator expression and comprehension execution. A new
+compile time AmbiguousTargetError is added to handle cases where it isn't
+clear to the compiler which target will be rebound by an augmented assignment.
+
+Finally, ``NAME := EXPR`` is proposed as a name rebinding expression that
+implements the new augmented assignment scoping rules, rather than implicitly
+defining a new local variable name.
+
+
+Background
+==========
+
+The case for allowing inline assignments at all is made in PEP 572. This
+PEP was initially going to propose an alternate surface syntax
+(``EXPR given NAME = EXPR``), while retaining the expression semantics from
+PEP 572, but that changed when reconsidering one of the initial motivating use
+cases for allowing embedded assignments at all: making it possible to calculate
+cumulative sums in comprehensions and generator expressions.
+
+An arguably clear spelling for that operation would be::
+
+    def cumulative_sums(data, start=0)
+        total = start
+        yield from (total += value for value in data)
+        return total
+
+Since that behaviour can't be readily explained in terms of any of the current
+name binding operations, a new name rebinding operator would need to be
+introduced, allowing the above to be described as equivalent to::
+
+    def cumulative_sums(data, start=0)
+        total = start
+        yield from (total := operator.iadd(total, value) for value in data)
+        return total
+
+The proposal in this PEP would allow the above code to work, while also allowing
+inline augmented assignment operations like the following::
+
+    remaining_attempts = 10
+    while remaining_attempts -= 1:
+        ... # Attempt the operation once
+
+With a preceding local name binding or declaration, it would allow name binding
+in if/elif chains::
+
+    m = pattern.match(data)
```

Comment:
FWIW, my initial inclination (based on this example) is to combine these two lines as-is (without any preceding locals-binding):

    if m := pattern.match(data):
        ...

It may be worth adding a comment to the example to reassure readers that there's a reason the code is the way it is:

    # We explicitly bind locally to avoid ambiguity (see the scope resolution rules below).
    m = pattern.match(data)
    if m:
        ...

This isn't that big a deal, but I figured I'd point it out so you can weigh my experience against the cost of extra text. :)